tests/Replication2/ReplicatedState/DocumentStateMachineTest.cpp

```diff
@@ -42,290 +43,416 @@ using namespace arangodb::replication2::test;
-    }
-    return nullptr;
+struct DocumentStateMachineTest : test::ReplicatedLogTest {
+  DocumentStateMachineTest() {
+    fakeit::Fake(Dtor(handlersFactoryMock));
+    fakeit::Fake(Dtor(agencyHandlerMock));
+    fakeit::Fake(Dtor(shardHandlerMock));
+    fakeit::Fake(Dtor(transactionMock));
+    fakeit::Fake(Dtor(dbGuardMock));
+
+    feature->registerStateType<DocumentState>(
+        std::string{DocumentState::NAME},
+        std::shared_ptr<IDocumentStateHandlersFactory>(
+            &handlersFactoryMock.get()));
   }
 
-  std::string database;
-  std::unordered_map<TransactionId,
-                     std::shared_ptr<MockDocumentStateTransaction>>
-      transactions;
-};
-
-struct MockDocumentStateHandlersFactory : IDocumentStateHandlersFactory {
-  MockDocumentStateHandlersFactory(
-      std::shared_ptr<IDocumentStateAgencyHandler> ah,
-      std::shared_ptr<IDocumentStateShardHandler> sh,
-      MockDocumentStateTransactionHandler*& transactionHandler)
-      : agencyHandler(std::move(ah)),
-        shardHandler(std::move(sh)),
-        transactionHandler(transactionHandler) {}
-
-  auto createAgencyHandler(GlobalLogIdentifier gid)
-      -> std::shared_ptr<IDocumentStateAgencyHandler> override {
-    return agencyHandler;
-  }
+  std::shared_ptr<ReplicatedStateFeature> feature =
+      std::make_shared<ReplicatedStateFeature>();
 
-  auto createShardHandler(GlobalLogIdentifier gid)
-      -> std::shared_ptr<IDocumentStateShardHandler> override {
-    return shardHandler;
+  fakeit::Mock<IDocumentStateHandlersFactory> handlersFactoryMock;
+  fakeit::Mock<IDocumentStateAgencyHandler> agencyHandlerMock;
+  fakeit::Mock<IDocumentStateShardHandler> shardHandlerMock;
+  fakeit::Mock<IDocumentStateTransaction> transactionMock;
+  fakeit::Mock<IDatabaseGuard> dbGuardMock;
+
+  void SetUp() override {
+    fakeit::When(Method(agencyHandlerMock, getCollectionPlan))
+        .AlwaysReturn(std::make_shared<VPackBuilder>());
+    fakeit::When(Method(agencyHandlerMock, reportShardInCurrent))
+        .AlwaysReturn(Result{});
+    fakeit::When(Method(handlersFactoryMock, createAgencyHandler))
+        .AlwaysDo([&](GlobalLogIdentifier gid) {
+          return std::shared_ptr<IDocumentStateAgencyHandler>(
+              &agencyHandlerMock.get());
+        });
+    fakeit::When(Method(handlersFactoryMock, createShardHandler))
+        .AlwaysDo([&](GlobalLogIdentifier gid) {
+          fakeit::When(Method(shardHandlerMock, createLocalShard))
+              .AlwaysReturn(ResultT<std::string>::success(
+                  DocumentStateShardHandler::stateIdToShardId(gid.id)));
+          return std::shared_ptr<IDocumentStateShardHandler>(
+              &shardHandlerMock.get());
+        });
+    fakeit::When(Method(handlersFactoryMock, createTransaction))
+        .AlwaysDo([&](DocumentLogEntry const&, IDatabaseGuard const&) {
+          return std::shared_ptr<IDocumentStateTransaction>(
+              &transactionMock.get());
+        });
```

Comment:
Do we really want to double down on fakeit? Is it that hard to write a mock without it? 
I consider fakeit to be an ugly hack. See [Limitations](https://github.com/eranpeer/FakeIt#limitations).
In particular stuff like `-O2` is not supported.

Reply:
We switched to gmock.